### PR TITLE
Add more retries for MPD and validate content length of chunk

### PR DESCRIPTION
**Pull request by smurching**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/smurching/mlflow/pull/10113?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10113/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10113
```

</p>
</details>

### Related Issues/PRs

 #xxx

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/utils/request_utils.py [hunk 1]
```diff
@@ -49,6 +49,16 @@ def download_chunk(*, range_start, range_end, headers, download_path, http_uri):
...
+        expected_length = response.headers.get("Content-Length")
+        if expected_length is not None:
+            actual_length = response.raw.tell()
+            expected_length = int(expected_length)
+            if actual_length < expected_length:
+                raise IOError(
+                    "incomplete read ({} bytes read, {} more expected)".format(
+                        actual_length, expected_length - actual_length
+                    )
+                )
```
Comment (harupy) [inline mlflow/utils/request_utils.py:61]: /ai give me python code to test this code

Comment (harupy) [reply mlflow/utils/request_utils.py:61]: Thanks!

File: mlflow/store/artifact/cloud_artifact_repo.py [hunk 2]
```diff
@@ -212,20 +213,32 @@ def _parallelized_download_from_cloud(self, file_size, remote_file_path, local_p
             )
 
             if failed_downloads:
```
Comment (harupy) [inline mlflow/store/artifact/cloud_artifact_repo.py:215]: We might want to throw immediately if the size of `failed_downloads` is large. Let's say you have 100 chunks to upload, and 90 of them fail, we throw.

Comment (BenWilson2) [reply mlflow/store/artifact/cloud_artifact_repo.py:215]: This might be a bit tricky to get the logic just right. If we do percentage basis, 2 failures out of 3 total would be the same as 200 failures out of 300 total (we would probably want to retry the first one, but not the 2nd one). If we do a scalar cutoff of say 50 (purely arbitrary for argument's sake), 49 failures out of 51 total is probably something we wouldn't want to retry, but failing 51 out of 5000 is probably something worth retrying. 
I'm worried about how complicated this logic might need to be to capture refined retry logic.

Do you think for now we could just brute-force retry all failures and revisit more advanced logic if we find it is an issue for users later?

File: mlflow/store/artifact/cloud_artifact_repo.py [hunk 3]
```diff
@@ -212,20 +213,30 @@ def _parallelized_download_from_cloud(self, file_size, remote_file_path, local_p
...
-                new_cloud_creds = self._get_read_credential_infos([remote_file_path])[0]
-                new_signed_uri = new_cloud_creds.signed_uri
-                new_headers = self._extract_headers_from_credentials(new_cloud_creds.headers)
+                num_retries = 3
```
Comment (BenWilson2) [inline mlflow/store/artifact/cloud_artifact_repo.py:216]: @harupy do you think it might be helpful to make this an environment configuration in case we see some adverse behavior? 

The scenario I'm thinking about is: what if we find in another few weeks that there is some failure mechanism for partial download in R2 that tends to resolve 95% of the time after 5 retries, but the success distribution falls to 50% after 3 retries. Or some similar behavior like this. 
Conversely, what if we find that retrying more than once in AWS or Azure is a complete waste of time and some users might want to fail fast if there are connection issues? 

What do you think about a config here that's defaulted to 3?

File: mlflow/store/artifact/cloud_artifact_repo.py [hunk 4]
```diff
@@ -212,20 +213,30 @@ def _parallelized_download_from_cloud(self, file_size, remote_file_path, local_p
...
+                                http_uri=new_signed_uri,
+                            )
+                            _logger.warning(
+                                f"Successfully downloaded chunk {chunk} for {remote_file_path}"
+                            )
+                            break
+                        except Exception:
+                            if retry == num_retries - 1:
+                                raise
+                        time.sleep(1)
```
Comment (BenWilson2) [inline mlflow/store/artifact/cloud_artifact_repo.py:239]: Similar to my comment above about retry count, what are your thoughts on a configurable sleep duration for retries? Since we're doing a loop over each partition transfer and only waiting 1s between each try, there may be considerations in the R2 provider where a longer wait may increase the probability of success on aggregate based on observations once this is released. 

Do you think a configurable value for this might be worthwhile?

File: mlflow/store/artifact/cloud_artifact_repo.py [hunk 5]
```diff
@@ -212,20 +213,30 @@ def _parallelized_download_from_cloud(self, file_size, remote_file_path, local_p
...
-                        download_path=local_path,
-                        http_uri=new_signed_uri,
-                    )
+                    new_cloud_creds = self._get_read_credential_infos([remote_file_path])[0]
+                    new_signed_uri = new_cloud_creds.signed_uri
+                    new_headers = self._extract_headers_from_credentials(new_cloud_creds.headers)
+                    _logger.warning(f"Retrying download of chunk {chunk} for {remote_file_path}")
+                    for retry in range(num_retries):
+                        try:
+                            _logger.warning(f"Attempt {retry + 1}/{num_retries}")
```
Comment (smurching) [inline mlflow/store/artifact/cloud_artifact_repo.py:224]:
```suggestion
                            _logger.debug(f"Attempt {retry + 1}/{num_retries}")
```

Drive-by comment, I was thinking when we ship this, should we use a lower severity log-level than `warning`? Maybe `debug` or `info`? Since retrying should be harmless/users shouldn't care most of the time

Comment (smurching) [reply mlflow/store/artifact/cloud_artifact_repo.py:224]: And especially for R2 it seems reasonably likely you'll need at least one part retry in the process of downloading a large model

Comment (BenWilson2) [reply mlflow/store/artifact/cloud_artifact_repo.py:224]: +1 good idea. Otherwise based on the rates that you were seeing during testing, this could get REALLY chatty on stdout / stderr. `debug` seems appropriate, matching retry policies that we see in Spark for failed fetches during DataFrame loads.

Comment (harupy) [reply mlflow/store/artifact/cloud_artifact_repo.py:224]: +1. Will make this change.

File: mlflow/store/artifact/cloud_artifact_repo.py (hunk 3 above)
Comment (harupy) [reply mlflow/store/artifact/cloud_artifact_repo.py:216]: private envs
```
_MLFLOW_MPD_NUM_RETRIES
_MLFLOW_MPD_RETRY_SLEEP_SECONDS
```

File: mlflow/store/artifact/cloud_artifact_repo.py (hunk 2 above)
Comment (harupy) [reply mlflow/store/artifact/cloud_artifact_repo.py:215]: revisit later sounds good.

File: mlflow/store/artifact/cloud_artifact_repo.py (hunk 5 above)
Comment (harupy) [reply mlflow/store/artifact/cloud_artifact_repo.py:224]: Chose info. debug is not printed out (unless you modify the logging level)

File: mlflow/utils/file_utils.py [hunk 6]
```diff
@@ -762,14 +771,40 @@ def run_download(chunk: _Chunk):
             chunk = futures[future]
             try:
                 future.result()
-            except Exception:
+            except Exception as e:
+                _logger.warning(
```
Comment (harupy) [inline mlflow/utils/file_utils.py:775]: the warning looks like this:

image

this might annoy users

File: mlflow/environment_variables.py [hunk 7]
```diff
@@ -440,3 +440,12 @@ def get(self):
...
 )
+
+# Private environment variable to specify the number of chunk download retries for multipart
+# download.
+_MLFLOW_MPD_NUM_RETRIES = _EnvironmentVariable("_MLFLOW_MPD_NUM_RETRIES", int, 3)
+# Private environment variable to specify the interval between chunk download retries for multipart
+# download.
+_MLFLOW_MPD_RETRY_INTERVAL_SECONDS = _EnvironmentVariable(
+    "_MLFLOW_MPD_RETRY_INTERVAL_SECONDS", int, 1
+)
```
Comment (harupy) [inline mlflow/environment_variables.py:451]: adding these for https://github.com/mlflow/mlflow/pull/10113#discussion_r1373305745

File: mlflow/utils/file_utils.py (hunk 6 above)
Comment (harupy) [reply mlflow/utils/file_utils.py:775]: If it does, we can use `debug`.

Comment (BenWilson2) [reply mlflow/utils/file_utils.py:775]: I think debug might be better. It still gives us the ability to troubleshoot a failure, but users don't typically see transfer retries in Spark. Might as well keep it consistent, I feel.

Comment (BenWilson2) [reply mlflow/utils/file_utils.py:775]: I don't feel this is a blocker, though.

**Review by BenWilson2 — APPROVED** [review]
LGTM! Thanks @harupy :)

File: mlflow/utils/file_utils.py (hunk 6 above)
Comment (BenWilson2) [reply mlflow/utils/file_utils.py:775]: I'll file a followup PR to change the logging level. Merging this PR.
